### ISSUE TITLE
Feature request: Set<TEntity>() method for IDbContextFactory<TContext>

### ISSUE THREAD

**Issue by wertzui** [dotnet/efcore#23113]
As `DbContext` is not thread save, you have to create multiple contexts when you want to execute multiple `async` queries. This is really hard to read because of the `using` blocks. It has become a little bit better with the new `using` statements, but is still feels like a lot to write.

```
using var context1 = _factory.CreateDbContext();
using var context2 = _factory.CreateDbContext();
using var context3 = _factory.CreateDbContext();

var result1Task = context1.Set<T1>().ToListAsync();
var result2Task = context2.Set<T2>().ToListAsync();
var result3Task = context3.Set<T3>().ToListAsync();

await Task.WhenAll(result1Task, result2Task, result3Task);

var result1 = result1Task.Result;
var result2 = result2Task.Result;
var result3 = result3Task.Result;
```

My request is to add a `Set<TEntity>()` method to `IDbContextFactory<TContext>` so you no longer have to write the using statements and are sure the contexts are disposed as soon as possible (not at the end of the method, but as soon as the query is executed).

```
var result1Task = _factory.Set<T1>().ToListAsync();
var result2Task =  _factory.Set<T2>().ToListAsync();
var result3Task =  _factory.Set<T3>().ToListAsync();

await Task.WhenAll(result1Task, result2Task, result3Task);

var result1 = result1Task.Result;
var result2 = result2Task.Result;
var result3 = result3Task.Result;
```

**Comment by roji** [comment]
The same disposal problem persists in your proposal - exactly when is the DbContext disposed? The new IDbContextFactory would need to return a set on an active, undisposed DbContext (which is allocated internally). In your example you immediately call ToListAsync on the returned DbSet, so it looks like the DbContext can be disposed internally. But the user can run several queries via the returned DbSet - so when/how is the context disposed?

**Comment by wertzui** [comment]
If the `Set<TEntity>` method on `IDbContextFactory<TContext>` would return `IQueryable<TEntity>` rather than `DbSet<TEntity>`, the context can be created and immediately disposed when the query is executed.

**Comment by roji** [comment]
Two small comments:

* This way of doing things is incompatible with streaming - you have to buffer all the results in a List rather than enumerate over the results. Streaming is a frequent reason why people want to do parallel queries in the first place.
* IQueryable wouldn't be fully functional, since some methods do require actual DbSet for some things (e.g. raw SQL with FromSql). This 

But more importantly... The same thing can be achieved pretty easily via an extension method over IDbContextFactory, which accepts a lambda, creates a DbContext and executes it. Something like the following:

```c#
static class DbContextFactoryExtensions
{
    public static async Task<List<TResult>> ToListAsync<TContext, TResult>(
        this IDbContextFactory<TContext> factory,
        Func<TContext, IQueryable<TResult>> query)
        where TContext : DbContext
    {
        await using var context = factory.CreateDbContext();
        return await query(context).ToListAsync();
    }
}

// Usage:
_ = await dbContextFactory.ToListAsync(ctx => ctx.Blogs.Where(b => b.Name == "foo"));

```

(thanks @smitpatel)

**Comment by wertzui** [comment]
> This way of doing things is incompatible with streaming - you have to buffer all the results in a List rather than enumerate over the results. Streaming is a frequent reason why people want to do parallel queries in the first place.

I came across a lot of scenarios where the query itself takes some time to execute on the server, but the result isn't that big that streaming would make a difference here (I'm mostly thinking about Web API calls here).

> IQueryable wouldn't be fully functional, since some methods do require actual DbSet for some things (e.g. raw SQL with FromSql). This

Looking at https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Query/Internal/FromSqlQueryRootExpression.cs I cannot see where it does need an actual DbSet.
The extension Method `FromSqlRaw` in https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Extensions/RelationalQueryableExtensions.cs needs the DbSet only to get the provider from it.
However using my suggested approach, the provider is not available while the query is constructed anyway (it has to be replaced/injected when the query is executed) so this would make no difference to a "normal" Linq query.

> But more importantly... The same thing can be achieved pretty easily via an extension method over IDbContextFactory, which accepts a lambda, creates a DbContext and executes it. Something like the following:

If you have something like this, but with a syntax that looks more like the "normal" way (`await factory.Set<Blog>().Where(b => b.Name == "foo").ToListAsync();`), I would be very happy 😄

**Comment by roji** [comment]
>> This way of doing things is incompatible with streaming - you have to buffer all the results in a List rather than enumerate over the results. Streaming is a frequent reason why people want to do parallel queries in the first place.
>
> I came across a lot of scenarios where the query itself takes some time to execute on the server, but the result isn't that big that streaming would make a difference here (I'm mostly thinking about Web API calls here).

Those scenarios definitely exist, but the opposite also exists, and we should avoid introducing an EF Core API which buffers without it being very clear.

>> IQueryable wouldn't be fully functional, since some methods do require actual DbSet for some things (e.g. raw SQL with FromSql).
>
> Looking at https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Query/Internal/FromSqlQueryRootExpression.cs I cannot see where it does need an actual DbSet. [...]

My point here wasn't about how FromSqlRaw is implemented - just that the method is defined over DbSet and not over IQueryable; so it wouldn't be usable following [your suggestion above](https://github.com/dotnet/efcore/issues/23113#issuecomment-717778082).

> If you have something like this, but with a syntax that looks more like the "normal" way (await factory.Set<Blog>().Where(b => b.Name == "foo").ToListAsync();), I would be very happy smile

To summarize:

1. This would work only for very specific scenario (single operations without any tracking/state)
2. It would probably lead people to unintentionally buffer when they could stream
3. The same thing can be achieved with only a slightly different construct via an extension method (as proposed above) - it seems to be mostly a matter of taste

I don't think this really belongs in EF Core... Of course, feel free to continue the discussion and raise other arguments or proposals.

**Comment by wertzui** [comment]
For anyone interested:
I started my own approach to get this working as an extension method to `IDbContextFactory`.
https://github.com/wertzui/EntityFrameworkCore.Parallel
